WalletWasabi/Tor/Socks5/TorTcpConnectionFactory.cs

```diff
@@ -19,82 +18,96 @@
 		{
 			TorSocks5EndPoint = endPoint;
-			TcpClient = new TcpClient(endPoint.AddressFamily);
-			AsyncLock = new AsyncLock();
 		}
 
-		#region PropertiesAndMembers
-
-		/// <summary>TCP connection to Tor's SOCKS5 server.</summary>
-		private TcpClient TcpClient { get; set; }
-
 		private EndPoint TorSocks5EndPoint { get; }
 
-		/// <summary>Transport stream for sending  HTTP/HTTPS requests through Tor's SOCKS5 server.</summary>
-		/// <remarks>This stream is not to be used to send commands to Tor's SOCKS5 server.</remarks>
-		private Stream Stream { get; set; }
+		/// <summary>
+		/// Creates a new connected TCP client connected to Tor SOCKS5 endpoint.
+		/// </summary>
+		/// <param name="host">Tor SOCKS5 host.</param>
+		/// <param name="port">Tor SOCKS5 port.</param>
+		/// <param name="useSsl">Whether to use SSL to send the HTTP request over Tor.</param>
+		/// <param name="isolateStream"><c>true</c> if a new Tor circuit is required for this HTTP request.</param>
+		/// <param name="cancellationToken">Cancellation token to cancel the asynchronous operation.</param>
+		/// <returns>New <see cref="TorTcpConnection"/> instance.</returns>
+		/// <exception cref="TorConnectionException">When <see cref="ConnectAsync(TcpClient, CancellationToken)"/> fails.</exception>
+		public async Task<TorTcpConnection> EstablishConnectionAsync(string host, int port, bool useSsl, bool isolateStream, CancellationToken cancellationToken = default)
+		{
+			TcpClient? tcpClient = null;
+			Stream? transportStream = null;
+
+			try
+			{
+				tcpClient = new TcpClient(TorSocks5EndPoint.AddressFamily);
 
-		private EndPoint RemoteEndPoint { get; set; }
+				transportStream = await ConnectAsync(tcpClient, cancellationToken).ConfigureAwait(false);
+				await HandshakeAsync(tcpClient, isolateStream, cancellationToken).ConfigureAwait(false);
+				await ConnectToDestinationAsync(tcpClient, host, port, cancellationToken).ConfigureAwait(false);
 
-		internal AsyncLock AsyncLock { get; }
+				if (useSsl)
+				{
+					transportStream = await UpgradeToSslAsync(tcpClient, host).ConfigureAwait(false);
+				}
 
-		#endregion PropertiesAndMembers
+				TorTcpConnection result = new(tcpClient, transportStream);
 
-		#region Initializers
+				transportStream = null;
+				tcpClient = null;
+				return result;
+			}
+			finally
+			{
+				transportStream?.Dispose();
+				tcpClient?.Dispose();
+			}
+		}
 
 		/// <summary>
 		/// Establishes TCP connection with Tor SOCKS5 endpoint.
 		/// </summary>
 		/// <exception cref="ArgumentException">This should never happen.</exception>
 		/// <exception cref="TorException">When connection to Tor SOCKS5 endpoint fails.</exception>
-		public async Task ConnectAsync(CancellationToken cancellationToken = default)
+		private async Task<NetworkStream> ConnectAsync(TcpClient tcpClient, CancellationToken cancellationToken = default)
 		{
-			using (await AsyncLock.LockAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
explanation: `AsyncLock` was removed because we do not need to hold a lock for creating a TCP connection to Tor SOCKS5 endpoint. You can create multiple connections to an endpoint in parallel if you want. That's common.

Note that `TorTcpConnectionFactory` is stateless.